File: packages/sdk/src/stream.ts
/**
 * WebSocket streaming API.
 *
 * Each method returns a {@link MelayaStream} that is both an async-iterable
 * (`for await (const frame of stream)`) and an event emitter
 * (`stream.on("message", cb)`). Uses the global `WebSocket` by default;
 * inject one (e.g. the `ws` package) for Node < 22.
 *
 * SECURITY / auth wire format: REST calls send the API key ONLY as an
 * `Authorization: Bearer mk_...` header (see `client.ts`). WebSocket
 * connections cannot carry custom headers from browsers, so the server
 * protocol differs per plane:
 *   - PUBLIC market-data streams authenticate with `?apiKey=` in the `wss://`
 *     URL (server protocol — redact it from proxy/APM/WebSocket query logs).
 *   - PRIVATE streams never expose the API key: a short-lived `wsTicket` is
 *     minted over REST first, and only `?wsTicket=` appears in the URL.
 */
import type { HttpClient } from "./client.js";
import type {
  SymbolQuery,
  OrderBookQuery,
  OhlcvQuery,
  TickerFrame,
  OrderBookFrame,
  TradeFrame,
  OhlcvFrame,
  LiquidationFrame,
} from "./types.js";

/** Minimal WebSocket surface shared by the browser, Node 22+ global, and `ws`. */
export interface WebSocketLike {
  addEventListener(type: string, listener: (ev: unknown) => void): void;
  close(): void;
  readyState: number;
}
export type WebSocketCtor = new (url: string) => WebSocketLike;

type StreamEvent = "message" | "open" | "close" | "error";

/**
 * A live stream of normalized frames. Iterate it, or attach listeners.
 *
 * Frames are buffered for the async-iterator only once iteration has started
 * (and the buffer is bounded), so listener-only consumers
 * (`stream.on("message", cb)`) never accumulate frames in memory.
 *
 * @example
 * ```ts
 * const s = melaya.stream.ticker({ exchange: "binance", symbol: "BTC/USDT", market: "spot" });
 * for await (const t of s) console.log(t.last);
 * ```
 */
export class MelayaStream<T> implements AsyncIterable<T> {
  /** Max frames buffered for a slow async-iterator before oldest are dropped. */
  private static readonly MAX_BUFFER = 10_000;

  private ws: WebSocketLike;
  private buffer: T[] = [];
  private waiting: ((r: IteratorResult<T>) => void) | null = null;
  private closed = false;
  /** True once `[Symbol.asyncIterator]()` has been called — gates buffering. */
  private iterating = false;
  private listeners: Record<StreamEvent, Array<(arg: unknown) => void>> = {
    message: [], open: [], close: [], error: [],
  };

  constructor(url: string, WebSocketImpl: WebSocketCtor) {
    this.ws = new WebSocketImpl(url);
    this.ws.addEventListener("open", (ev) => this.emit("open", ev));
    this.ws.addEventListener("error", (ev) => this.emit("error", ev));
    this.ws.addEventListener("close", () => this.finish());
    this.ws.addEventListener("message", (ev) => {
      const raw = (ev as { data?: unknown }).data;
      let frame: T;
      try {
        const text = typeof raw === "string" ? raw : raw instanceof ArrayBuffer ? new TextDecoder().decode(raw) : String(raw);
        frame = JSON.parse(text) as T;
      } catch {
        return; // ignore non-JSON keep-alive frames
      }
      this.emit("message", frame);
      if (this.waiting) {
        const w = this.waiting;
        this.waiting = null;
        w({ value: frame, done: false });
      } else if (this.iterating) {
        // Only buffer for the async-iterator once iteration has started, so
        // listener-only consumers don't grow memory unboundedly.
        this.buffer.push(frame);
        if (this.buffer.length > MelayaStream.MAX_BUFFER) this.buffer.shift();
      }
    });
  }

  /** Subscribe to stream events. Returns an unsubscribe function. */
  on(event: "message", cb: (frame: T) => void): () => void;
  on(event: "open" | "close" | "error", cb: (ev: unknown) => void): () => void;
  on(event: StreamEvent, cb: (arg: never) => void): () => void {
    this.listeners[event].push(cb as (arg: unknown) => void);
    return () => {
      this.listeners[event] = this.listeners[event].filter((l) => l !== cb);
    };
  }

  /** Close the underlying socket and end iteration. */
  close(): void {
    try { this.ws.close(); } catch { /* already closing */ }
    this.finish();
  }

  private emit(event: StreamEvent, arg: unknown): void {
    for (const l of this.listeners[event]) l(arg);
  }

  private finish(): void {
    if (this.closed) return;
    this.closed = true;
    this.emit("close", undefined);
    if (this.waiting) {
      const w = this.waiting;
      this.waiting = null;
      w({ value: undefined as unknown as T, done: true });
    }
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    this.iterating = true;
    return {
      next: (): Promise<IteratorResult<T>> => {
        if (this.buffer.length > 0) {
          return Promise.resolve({ value: this.buffer.shift() as T, done: false });
        }
        if (this.closed) {
          return Promise.resolve({ value: undefined as unknown as T, done: true });
        }
        return new Promise<IteratorResult<T>>((resolve) => {
          this.waiting = resolve;
        });
      },
      return: (): Promise<IteratorResult<T>> => {
        this.close();
        return Promise.resolve({ value: undefined as unknown as T, done: true });
      },
    };
  }
}

export class StreamAPI {
  constructor(
    private readonly apiKey: string,
    private readonly wsUrl: string,
    private readonly WebSocketImpl: WebSocketCtor | undefined,
    private readonly http: HttpClient,
  ) {}

  /** Live ticker frames (fires only when the normalized ticker advances). */
  ticker(q: SymbolQuery): MelayaStream<TickerFrame> {
    return this.open<TickerFrame>("/ws/ticker", { ...q });
  }

  /** Live order-book frames. */
  orderbook(q: OrderBookQuery): MelayaStream<OrderBookFrame> {
    return this.open<OrderBookFrame>("/ws/orderbook", { ...q });
  }

  /** Live OHLCV candle frames. */
  ohlcv(q: OhlcvQuery): MelayaStream<OhlcvFrame> {
    return this.open<OhlcvFrame>("/ws/ohlcv", { ...q });
  }

  /** Live public-trade frames. */
  trades(q: SymbolQuery): MelayaStream<TradeFrame> {
    return this.open<TradeFrame>("/ws/public-trades", { ...q });
  }

  /** Cross-exchange liquidation firehose. Omit `exchange` for all venues. */
  liquidations(q: { exchange?: string } = {}): MelayaStream<LiquidationFrame> {
    return this.open<LiquidationFrame>("/ws/liquidations", { ...q });
  }

  // ── Private feeds (authenticated; ticket-minted) ──────────────────────────

  /**
   * Live strategy events for your account: cycle markers, agent messages,
   * approval requests, executions, and status changes across every strategy
   * you own. Mints a short-lived ticket, then opens `/ws/strategies`.
   *
   * @example
   * ```ts
   * const s = await melaya.stream.strategies();
   * for await (const ev of s) console.log(ev.type, ev.strategyId);
   * ```
   */
  async strategies(): Promise<MelayaStream<Record<string, unknown>>> {
    return this.openPrivate<Record<string, unknown>>("/ws/strategies", "strategies", {});
  }

  /**
   * Live private account feed for one connected exchange key (balance,
   * positions, your orders/fills). Pass the `apiKeyId` from
   * `melaya.account.keys()`. Mints a key-scoped ticket, then opens `/ws/private`.
   */
  async private(q: { exchange: string; market?: string; apiKeyId?: string; keyId?: string; symbol?: string }): Promise<MelayaStream<Record<string, unknown>>> {
    return this.openPrivate<Record<string, unknown>>("/ws/private", "private", {
      exchange: q.exchange,
      market: q.market,
      apiKeyId: q.apiKeyId,
      keyId: q.keyId,
      symbol: q.symbol,
    });
  }

  /** Mint a `wsTicket` for the given private stream, then open the socket with it. */
  private async openPrivate<T>(path: string, stream: string, body: Record<string, unknown>): Promise<MelayaStream<T>> {
    if (!this.WebSocketImpl) {
      throw new Error("Melaya: no global `WebSocket` found. Pass `WebSocket` in options (Node < 22, e.g. the `ws` package).");
    }
    const clean: Record<string, unknown> = { stream };
    for (const [k, v] of Object.entries(body)) if (v !== undefined && v !== null) clean[k] = v;
    const { wsTicket } = await this.http.post<{ wsTicket: string }>("/api/v1/private/private-ticket", clean);
    const u = new URL(path.replace(/^\//, ""), this.wsUrl.endsWith("/") ? this.wsUrl : this.wsUrl + "/");
    u.searchParams.set("wsTicket", wsTicket);
    return new MelayaStream<T>(u.toString(), this.WebSocketImpl);
  }

  private open<T>(path: string, params: Record<string, string | number | undefined>): MelayaStream<T> {
    if (!this.apiKey) {
      throw new Error("Melaya: public market streams require an `apiKey`; session-token clients can use platform events.");
    }
    if (!this.WebSocketImpl) {
      throw new Error("Melaya: no global `WebSocket` found. Pass `WebSocket` in options (Node < 22, e.g. the `ws` package).");
    }
    const u = new URL(path.replace(/^\//, ""), this.wsUrl.endsWith("/") ? this.wsUrl : this.wsUrl + "/");
    // Public-stream server protocol: the key rides as `?apiKey=` on the wss
    // URL (WebSocket clients cannot set headers). REST never does this.
    u.searchParams.set("apiKey", this.apiKey);
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null) u.searchParams.set(k, String(v));
    }
    return new MelayaStream<T>(u.toString(), this.WebSocketImpl);
  }
}

File: packages/sdk-go/melaya/stream.go
// Stream API — WebSocket streaming (public market data + private feeds).
package melaya

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"sync"

	"github.com/gorilla/websocket"
)

// Frame is a generic JSON frame from a WebSocket stream.
type Frame map[string]interface{}

// Stream is a live channel of JSON frames from a Melaya WebSocket.
// Read from the Ch channel; close with Close().
type Stream struct {
	Ch     <-chan Frame
	Errors <-chan error
	close  func()
}

// Close shuts down the underlying WebSocket connection.
func (s *Stream) Close() {
	if s.close != nil {
		s.close()
	}
}

// StreamAPI manages WebSocket connections to Melaya.
type StreamAPI struct {
	apiKey      string
	wsURL       string
	h           *httpClient
	insecureTLS bool
}

func (s *StreamAPI) dialer() *websocket.Dialer {
	d := *websocket.DefaultDialer
	if s.insecureTLS {
		d.TLSClientConfig = &tls.Config{InsecureSkipVerify: true} //nolint:gosec
	}
	return &d
}

func (s *StreamAPI) open(path string, params map[string]string) (*Stream, error) {
	u, err := url.Parse(s.wsURL + path)
	if err != nil {
		return nil, err
	}
	q := u.Query()
	q.Set("apiKey", s.apiKey)
	for k, v := range params {
		if v != "" {
			q.Set(k, v)
		}
	}
	u.RawQuery = q.Encode()

	conn, _, err := s.dialer().Dial(u.String(), http.Header{"Authorization": []string{"Bearer " + s.apiKey}})
	if err != nil {
		return nil, fmt.Errorf("melaya stream: dial %s: %w", path, err)
	}
	return newStream(conn), nil
}

func (s *StreamAPI) openWithTicket(path, stream string, extra map[string]interface{}) (*Stream, error) {
	body := map[string]interface{}{"stream": stream}
	for k, v := range extra {
		if v == nil {
			continue
		}
		if str, ok := v.(string); ok && str == "" {
			continue
		}
		body[k] = v
	}
	ctx := context.Background()
	data, err := s.h.post(ctx, "/api/v1/private/private-ticket", body)
	if err != nil {
		return nil, fmt.Errorf("melaya stream: mint ticket for %s: %w", stream, err)
	}
	var env struct {
		WsTicket string `json:"wsTicket"`
	}
	if err := json.Unmarshal(data, &env); err != nil || env.WsTicket == "" {
		return nil, fmt.Errorf("melaya stream: no wsTicket in response")
	}

	u, err := url.Parse(s.wsURL + path)
	if err != nil {
		return nil, err
	}
	q := u.Query()
	q.Set("wsTicket", env.WsTicket)
	u.RawQuery = q.Encode()

	conn, _, err := s.dialer().Dial(u.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("melaya stream: dial %s: %w", path, err)
	}
	return newStream(conn), nil
}

func newStream(conn *websocket.Conn) *Stream {
	ch := make(chan Frame, 64)
	errCh := make(chan error, 1)
	done := make(chan struct{})

	go func() {
		defer close(ch)
		defer close(errCh)
		for {
			select {
			case <-done:
				return
			default:
			}
			_, msg, err := conn.ReadMessage()
			if err != nil {
				select {
				case errCh <- err:
				default:
				}
				return
			}
			var f Frame
			if err := json.Unmarshal(msg, &f); err != nil {
				continue // ignore non-JSON keep-alive frames
			}
			select {
			case ch <- f:
			case <-done:
				return
			}
		}
	}()

	var once sync.Once
	return &Stream{
		Ch:     ch,
		Errors: errCh,
		close: func() {
			once.Do(func() {
				close(done)
				conn.Close()
			})
		},
	}
}

// ── Public streams ─────────────────────────────────────────────────────────────

// Ticker opens a live ticker stream for one symbol.
func (s *StreamAPI) Ticker(exchange, symbol, market string) (*Stream, error) {
	return s.open("/ws/ticker", map[string]string{
		"exchange": exchange,
		"symbol":   symbol,
		"market":   market,
	})
}

// Orderbook opens a live order-book stream.
func (s *StreamAPI) Orderbook(exchange, symbol, market string, limit int) (*Stream, error) {
	params := map[string]string{"exchange": exchange, "symbol": symbol, "market": market}
	if limit > 0 {
		params["limit"] = fmt.Sprintf("%d", limit)
	}
	return s.open("/ws/orderbook", params)
}

// Ohlcv opens a live OHLCV candle stream.
func (s *StreamAPI) Ohlcv(exchange, symbol, timeframe, market string) (*Stream, error) {
	return s.open("/ws/ohlcv", map[string]string{
		"exchange":  exchange,
		"symbol":    symbol,
		"timeframe": timeframe,
		"market":    market,
	})
}

// Trades opens a live public-trades stream.
func (s *StreamAPI) Trades(exchange, symbol, market string) (*Stream, error) {
	return s.open("/ws/public-trades", map[string]string{
		"exchange": exchange,
		"symbol":   symbol,
		"market":   market,
	})
}

// Liquidations opens the cross-exchange liquidation firehose.
func (s *StreamAPI) Liquidations(exchange string) (*Stream, error) {
	return s.open("/ws/liquidations", map[string]string{"exchange": exchange})
}

// ── Private streams (ticket-minted) ──────────────────────────────────────────

// Strategies opens the live strategy-events stream for your account.
func (s *StreamAPI) Strategies() (*Stream, error) {
	return s.openWithTicket("/ws/strategies", "strategies", nil)
}

// Private opens the live private account feed for one connected exchange key.
func (s *StreamAPI) Private(exchange, market, apiKeyID, keyID, symbol string) (*Stream, error) {
	extra := map[string]interface{}{
		"exchange": exchange,
		"market":   market,
		"apiKeyId": apiKeyID,
		"keyId":    keyID,
		"symbol":   symbol,
	}
	return s.openWithTicket("/ws/private", "private", extra)
}
